src/ILCompiler.ReadyToRun/src/JitInterface/CorInfoImpl.ReadyToRun.cs

```diff
@@ -221,7 +214,7 @@ private bool getReadyToRunHelper(ref CORINFO_RESOLVED_TOKEN pResolvedToken, ref
                         Debug.Assert(typeToInitialize.IsCanonicalSubtype(CanonicalFormKind.Any));
 
                         DefType helperArg = typeToInitialize.ConvertToSharedRuntimeDeterminedForm();
-                        MethodContext methodContext = new MethodContext(methodFromContext(pResolvedToken.tokenContext), new IntPtr(pResolvedToken.tokenContext));
+                        MethodContext methodContext = new MethodContext(methodFromContext(pResolvedToken.tokenContext));
```

Comment:
`methodFromContext` will return NULL when the context is type. It means you will get the bogus sharing when the context is type again.

I think you should add a method like this:
```
private TypeSystemEntity entityFromContext(CORINFO_CONTEXT_STRUCT* contextStruct)
{
    return HandleToObject((CORINFO_CLASS_STRUCT_*)((ulong)contextStruct & ~(ulong)CorInfoContextFlags.CORINFO_CONTEXTFLAGS_MASK));
}
```

And use this value as the MethodContext for generic dictionary lookups.

Reply:
Thanks for the suggestions, however I'm somewhat confused; after looking at the related methods method / typeFromContext, I'm wondering whether it should rather be

<code>
        private TypeSystemEntity entityFromContext(CORINFO_CONTEXT_STRUCT* contextStruct)
        {
            if (((ulong)contextStruct & (ulong)CorInfoContextFlags.CORINFO_CONTEXTFLAGS_MASK) == (ulong)CorInfoContextFlags.CORINFO_CONTEXTFLAGS_CLASS)
            {
                return HandleToObject((CORINFO_CLASS_STRUCT_*)((ulong)contextStruct & ~(ulong)CorInfoContextFlags.CORINFO_CONTEXTFLAGS_MASK));
            }
            else
            {
                return HandleToObject((CORINFO_METHOD_STRUCT_*)((ulong)contextStruct & ~(ulong)CorInfoContextFlags.CORINFO_CONTEXTFLAGS_MASK));
            }
        }
</code>

Could you please enlighten me as to whether I'm completely off and / or whether we're fine ignoring the distinction between CORINFO_CLASS_STRUCT and CORINFO_METHOD_STRUCT?

Reply:
Fixed in 2nd commit. For now I used my conservative interpretation of your suggestion per my previous response.

Reply:
Yes, your implementation is correct as well.

What I had in mind was to use the HandleToObject overload that returns object that should work as well. Here is the corrected version of my original suggestion:

```
private TypeSystemEntity entityFromContext(CORINFO_CONTEXT_STRUCT* contextStruct)
{
    return (TypeSystemEntity)HandleToObject((IntPtr)((ulong)contextStruct & ~(ulong)CorInfoContextFlags.CORINFO_CONTEXTFLAGS_MASK));
}
```

I will leave it up to you to pick which one you like better.

Reply:
OK, I agree this is definitely simpler, I previously didn't fully appreciate that both share the same map. Fixed according to your suggestion in 3rd commit.